# Project: Investigate Medical Dataset for No Show Appointments

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
<li><a href="#eda">Exploratory Data Analysis</a></li>
<li><a href="#conclusions">Conclusions</a></li>
</ul>

<a id='intro'></a>
## Introduction

This dataset collects information from 100k medical appointments in Brazil and is focused on the question of whether or not patients show up
for their appointment.

### Questions
1. What factors are important for us to know in order to predict if a patient will show up for their scheduled appointment?

In [258]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime
%matplotlib inline

<a id='wrangling'></a>
## Data Wrangling

### Load data

In [259]:
df = pd.read_csv('noshow_data.csv', encoding = "utf-8")

### Basic information about Data

In [260]:
# no of samples and columns in the dataset
df.shape

(110527, 14)

In [261]:
# Datatypes of columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110527 entries, 0 to 110526
Data columns (total 14 columns):
PatientId         110527 non-null float64
AppointmentID     110527 non-null int64
Gender            110527 non-null object
ScheduledDay      110527 non-null object
AppointmentDay    110527 non-null object
Age               110527 non-null int64
Neighbourhood     110527 non-null object
Scholarship       110527 non-null int64
Hipertension      110527 non-null int64
Diabetes          110527 non-null int64
Alcoholism        110527 non-null int64
Handcap           110527 non-null int64
SMS_received      110527 non-null int64
No-show           110527 non-null object
dtypes: float64(1), int64(8), object(5)
memory usage: 11.8+ MB


In [262]:
# initial look at the data
df.head(5)

PatientId  AppointmentID Gender          ScheduledDay  \
0  2.987250e+13        5642903      F  2016-04-29T18:38:08Z   
1  5.589978e+14        5642503      M  2016-04-29T16:08:27Z   
2  4.262962e+12        5642549      F  2016-04-29T16:19:04Z   
3  8.679512e+11        5642828      F  2016-04-29T17:29:31Z   
4  8.841186e+12        5642494      F  2016-04-29T16:07:23Z   

         AppointmentDay  Age      Neighbourhood  Scholarship  Hipertension  \
0  2016-04-29T00:00:00Z   62    JARDIM DA PENHA            0             1   
1  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             0   
2  2016-04-29T00:00:00Z   62      MATA DA PRAIA            0             0   
3  2016-04-29T00:00:00Z    8  PONTAL DE CAMBURI            0             0   
4  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             1   

   Diabetes  Alcoholism  Handcap  SMS_received No-show  
0         0           0        0             0      No  
1         0           0        0             0      No  
2         0           0        0             0      No  
3         0           0        0             0      No  
4         1           0        0             0      No

In [263]:
# finding duplicate rows in the dataset
df.duplicated().sum()

0

#### No duplicated row found in data

In [264]:
# gather descriptive statistics about data
df.describe()

PatientId  AppointmentID            Age    Scholarship  \
count  1.105270e+05   1.105270e+05  110527.000000  110527.000000   
mean   1.474963e+14   5.675305e+06      37.088874       0.098266   
std    2.560949e+14   7.129575e+04      23.110205       0.297675   
min    3.921784e+04   5.030230e+06      -1.000000       0.000000   
25%    4.172614e+12   5.640286e+06      18.000000       0.000000   
50%    3.173184e+13   5.680573e+06      37.000000       0.000000   
75%    9.439172e+13   5.725524e+06      55.000000       0.000000   
max    9.999816e+14   5.790484e+06     115.000000       1.000000   

        Hipertension       Diabetes     Alcoholism        Handcap  \
count  110527.000000  110527.000000  110527.000000  110527.000000   
mean        0.197246       0.071865       0.030400       0.022248   
std         0.397921       0.258265       0.171686       0.161543   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       4.000000   

        SMS_received  
count  110527.000000  
mean        0.321026  
std         0.466873  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         1.000000  
max         1.000000

In [265]:
# looking at unique count in columns
df.nunique()

PatientId          62299
AppointmentID     110527
Gender                 2
ScheduledDay      103549
AppointmentDay        27
Age                  104
Neighbourhood         81
Scholarship            2
Hipertension           2
Diabetes               2
Alcoholism             2
Handcap                5
SMS_received           2
No-show                2
dtype: int64

In [266]:
df['Age'].unique()

array([ 62,  56,   8,  76,  23,  39,  21,  19,  30,  29,  22,  28,  54,
        15,  50,  40,  46,   4,  13,  65,  45,  51,  32,  12,  61,  38,
        79,  18,  63,  64,  85,  59,  55,  71,  49,  78,  31,  58,  27,
         6,   2,  11,   7,   0,   3,   1,  69,  68,  60,  67,  36,  10,
        35,  20,  26,  34,  33,  16,  42,   5,  47,  17,  41,  44,  37,
        24,  66,  77,  81,  70,  53,  75,  73,  52,  74,  43,  89,  57,
        14,   9,  48,  83,  72,  25,  80,  87,  88,  84,  82,  90,  94,
        86,  91,  98,  92,  96,  93,  95,  97, 102, 115, 100,  99,  -1])

#### Found -1 age that looks human error. As per the documentation discussion 0 is new born babies under 1 year old. Also, 115 looks outlier in data.

In [267]:
df['Neighbourhood'].unique()

array(['JARDIM DA PENHA', 'MATA DA PRAIA', 'PONTAL DE CAMBURI',
       'REPÚBLICA', 'GOIABEIRAS', 'ANDORINHAS', 'CONQUISTA',
       'NOVA PALESTINA', 'DA PENHA', 'TABUAZEIRO', 'BENTO FERREIRA',
       'SÃO PEDRO', 'SANTA MARTHA', 'SÃO CRISTÓVÃO', 'MARUÍPE',
       'GRANDE VITÓRIA', 'SÃO BENEDITO', 'ILHA DAS CAIEIRAS',
       'SANTO ANDRÉ', 'SOLON BORGES', 'BONFIM', 'JARDIM CAMBURI',
       'MARIA ORTIZ', 'JABOUR', 'ANTÔNIO HONÓRIO', 'RESISTÊNCIA',
       'ILHA DE SANTA MARIA', 'JUCUTUQUARA', 'MONTE BELO',
       'MÁRIO CYPRESTE', 'SANTO ANTÔNIO', 'BELA VISTA', 'PRAIA DO SUÁ',
       'SANTA HELENA', 'ITARARÉ', 'INHANGUETÁ', 'UNIVERSITÁRIO',
       'SÃO JOSÉ', 'REDENÇÃO', 'SANTA CLARA', 'CENTRO', 'PARQUE MOSCOSO',
       'DO MOSCOSO', 'SANTOS DUMONT', 'CARATOÍRA', 'ARIOVALDO FAVALESSA',
       'ILHA DO FRADE', 'GURIGICA', 'JOANA D´ARC', 'CONSOLAÇÃO',
       'PRAIA DO CANTO', 'BOA VISTA', 'MORADA DE CAMBURI', 'SANTA LUÍZA',
       'SANTA LÚCIA', 'BARRO VERMELHO', 'ESTRELINHA', 'FORTE SÃO 

#### Some of the hospital names contains special character but we don't need to treat them specially for our analysis

In [268]:
print("Gender: {} Scholarship: {}".format(df['Gender'].unique(), df['Scholarship'].unique()))

Gender: ['F' 'M'] Scholarship: [0 1]


In [269]:
print("Hipertension: {} Diabetes: {}".format(df['Hipertension'].unique(), df['Diabetes'].unique()))

Hipertension: [1 0] Diabetes: [0 1]


In [270]:
print("Alcoholism: {} Handcap: {}".format(df['Alcoholism'].unique(), df['Handcap'].unique()))

Alcoholism: [0 1] Handcap: [0 1 2 3 4]


#### As per the original Kaggle documentation https://www.kaggle.com/joniarroba/noshowappointments Handcap has only 2 values True or False. Something to take care during data cleanup. 

In [271]:
print("SMS_received: {} No-show: {}".format(df['SMS_received'].unique(), df['No-show'].unique()))

SMS_received: [0 1] No-show: ['No' 'Yes']


### Observation about data
1. Few columns like Handcap, Hipertension and No-show are misspelled or not consistent with other column names
2. ScheduledDay and AppointmentDay stores Date and time but datatype is Object.
3. The timestamp in AppointmentDay is all set to 00:00:00 that looks irrelevant for analysis and needs to be removed. Further ScheduledDay timestamp is of no importance without  AppintmentDay and no relation could be established between both dates. Hance, ScheduledDay timestamp could be removed. 
4. PatientId and AppointmentID are not relevant to our analysis and needs to be dropped.
5. Age is -1 indicates there's human data error that needs further investigation.
6. Need to find any outliers present in ScheduledDay and AppointmentDay where AppointmentDay is before ScheduledDay.
7. Correct Handicap Data values

### Cleaning Column Labels

In [272]:
# rename Handcap to Handicap, Hipertension to Hypertension and No-show to No_show
df = df.rename(columns = {'Handcap': 'Handicap', 'Hipertension': 'Hypertension','No-show': 'No_show'})

In [273]:
# check the columns are renamed
df.head(1)

PatientId  AppointmentID Gender          ScheduledDay  \
0  2.987250e+13        5642903      F  2016-04-29T18:38:08Z   

         AppointmentDay  Age    Neighbourhood  Scholarship  Hypertension  \
0  2016-04-29T00:00:00Z   62  JARDIM DA PENHA            0             1   

   Diabetes  Alcoholism  Handicap  SMS_received No_show  
0         0           0         0             0      No

### Drop columns

In [274]:
# drop patientid and appointmentid columns which are of no significance for analysis
df.drop(['PatientId', 'AppointmentID'], axis=1, inplace=True)

In [275]:
# confirm the columns are dropped
df.head(1)

Gender          ScheduledDay        AppointmentDay  Age    Neighbourhood  \
0      F  2016-04-29T18:38:08Z  2016-04-29T00:00:00Z   62  JARDIM DA PENHA   

   Scholarship  Hypertension  Diabetes  Alcoholism  Handicap  SMS_received  \
0            0             1         0           0         0             0   

  No_show  
0      No

### Remove outliers

In [276]:
# find number of rows where age < 0
len(df.query("Age < 0")), len(df)

(1, 110527)

In [277]:
# choose age >= 0 and drop ouliers in data
df = df.query("Age >=0")

In [278]:
# check the outlier is removed and the count of samples
len(df.query("Age < 0")), len(df)

(0, 110526)

### Extract Date from ScheduledDay and AppointmentDay Columns

In [279]:
# define a function to extract date
def extract_date(d):
    return pd.to_datetime(d).date()

In [280]:
df['ScheduledDay'] = df['ScheduledDay'].apply(extract_date)

In [281]:
df['AppointmentDay'] = df['AppointmentDay'].apply(extract_date)

In [282]:
df.head(1)

Gender ScheduledDay AppointmentDay  Age    Neighbourhood  Scholarship  \
0      F   2016-04-29     2016-04-29   62  JARDIM DA PENHA            0   

   Hypertension  Diabetes  Alcoholism  Handicap  SMS_received No_show  
0             1         0           0         0             0      No

In [283]:
df.dtypes

Gender            object
ScheduledDay      object
AppointmentDay    object
Age                int64
Neighbourhood     object
Scholarship        int64
Hypertension       int64
Diabetes           int64
Alcoholism         int64
Handicap           int64
SMS_received       int64
No_show           object
dtype: object

In [284]:
df['ScheduledDay'] = pd.to_datetime(df['ScheduledDay'])

In [285]:
df['AppointmentDay'] = pd.to_datetime(df['AppointmentDay'])

In [286]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 110526 entries, 0 to 110526
Data columns (total 12 columns):
Gender            110526 non-null object
ScheduledDay      110526 non-null datetime64[ns]
AppointmentDay    110526 non-null datetime64[ns]
Age               110526 non-null int64
Neighbourhood     110526 non-null object
Scholarship       110526 non-null int64
Hypertension      110526 non-null int64
Diabetes          110526 non-null int64
Alcoholism        110526 non-null int64
Handicap          110526 non-null int64
SMS_received      110526 non-null int64
No_show           110526 non-null object
dtypes: datetime64[ns](2), int64(7), object(3)
memory usage: 11.0+ MB


In [287]:
df.head()

Gender ScheduledDay AppointmentDay  Age      Neighbourhood  Scholarship  \
0      F   2016-04-29     2016-04-29   62    JARDIM DA PENHA            0   
1      M   2016-04-29     2016-04-29   56    JARDIM DA PENHA            0   
2      F   2016-04-29     2016-04-29   62      MATA DA PRAIA            0   
3      F   2016-04-29     2016-04-29    8  PONTAL DE CAMBURI            0   
4      F   2016-04-29     2016-04-29   56    JARDIM DA PENHA            0   

   Hypertension  Diabetes  Alcoholism  Handicap  SMS_received No_show  
0             1         0           0         0             0      No  
1             0         0           0         0             0      No  
2             0         0           0         0             0      No  
3             0         0           0         0             0      No  
4             1         1           0         0             0      No

### Finding outlier in ScheduledDay and AppointmentDay Data

In [291]:
# check if scheduled date is after appointment date
df_outlier['outlier'] = (df['AppointmentDay'] - df['ScheduledDay']).dt.days

/Users/sarora/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: modifications to a property of a datetimelike object are not supported and are discarded. Change values on the original.
  


In [292]:
df_outlier['outlier'].min()

-6

In [295]:
outlier_ids = df[df_outlier['outlier'] < 0].index

In [297]:
outlier_ids

Int64Index([27033, 55226, 64175, 71533, 72362], dtype='int64')

In [299]:
# removing 6 rows as they're treated as outliers or human error
df.drop(outlier_ids, inplace=True)

In [300]:
df.shape

(110521, 12)

In [301]:
# count of each value in Handicap Data
df.Handicap.value_counts()

0    108282
1      2040
2       183
3        13
4         3
Name: Handicap, dtype: int64

### Since 0 has a count 108285, it's safe to assume other values [1,2,3,4] indicates the patient is Handicap. We need to convert  [1,2,3,4] to 1.

In [302]:
# converting [1,2,3,4] to 1 
df['Handicap'] = (df.Handicap > 0).astype(np.int)

In [303]:
df['Handicap']

0         0
1         0
2         0
3         0
4         0
         ..
110522    0
110523    0
110524    0
110525    0
110526    0
Name: Handicap, Length: 110521, dtype: int64

In [304]:
# count Handicap to confirm numbers
df.Handicap.value_counts()

0    108282
1      2239
Name: Handicap, dtype: int64

In [305]:
df.head()

Gender ScheduledDay AppointmentDay  Age      Neighbourhood  Scholarship  \
0      F   2016-04-29     2016-04-29   62    JARDIM DA PENHA            0   
1      M   2016-04-29     2016-04-29   56    JARDIM DA PENHA            0   
2      F   2016-04-29     2016-04-29   62      MATA DA PRAIA            0   
3      F   2016-04-29     2016-04-29    8  PONTAL DE CAMBURI            0   
4      F   2016-04-29     2016-04-29   56    JARDIM DA PENHA            0   

   Hypertension  Diabetes  Alcoholism  Handicap  SMS_received No_show  
0             1         0           0         0             0      No  
1             0         0           0         0             0      No  
2             0         0           0         0             0      No  
3             0         0           0         0             0      No  
4             1         1           0         0             0      No

### Feature Extraction

#### Waiting_Time in days

In [309]:
df['Waiting_Time'] = (df['AppointmentDay'] - df['ScheduledDay']).dt.days

In [310]:
df['Waiting_Time'].min(), df['Waiting_Time'].max()

(0, 179)

In [311]:
df.head()

Gender ScheduledDay AppointmentDay  Age      Neighbourhood  Scholarship  \
0      F   2016-04-29     2016-04-29   62    JARDIM DA PENHA            0   
1      M   2016-04-29     2016-04-29   56    JARDIM DA PENHA            0   
2      F   2016-04-29     2016-04-29   62      MATA DA PRAIA            0   
3      F   2016-04-29     2016-04-29    8  PONTAL DE CAMBURI            0   
4      F   2016-04-29     2016-04-29   56    JARDIM DA PENHA            0   

   Hypertension  Diabetes  Alcoholism  Handicap  SMS_received No_show  \
0             1         0           0         0             0      No   
1             0         0           0         0             0      No   
2             0         0           0         0             0      No   
3             0         0           0         0             0      No   
4             1         1           0         0             0      No   

   waiting_time  Waiting_Time  
0             0             0  
1             0             0  
2             0             0  
3             0             0  
4             0             0

#### Weekday

In [312]:
mapDayOfWeek={0:'Monday', 1:'Tuesday', 2:'Wednesday', 3:'Thursday', 4:'Friday', 5:'Saturday', 6:'Sunday'}
df['Weekday'] = df['AppointmentDay'].dt.dayofweek.map(mapDayOfWeek)

In [313]:
df.head()

Gender ScheduledDay AppointmentDay  Age      Neighbourhood  Scholarship  \
0      F   2016-04-29     2016-04-29   62    JARDIM DA PENHA            0   
1      M   2016-04-29     2016-04-29   56    JARDIM DA PENHA            0   
2      F   2016-04-29     2016-04-29   62      MATA DA PRAIA            0   
3      F   2016-04-29     2016-04-29    8  PONTAL DE CAMBURI            0   
4      F   2016-04-29     2016-04-29   56    JARDIM DA PENHA            0   

   Hypertension  Diabetes  Alcoholism  Handicap  SMS_received No_show  \
0             1         0           0         0             0      No   
1             0         0           0         0             0      No   
2             0         0           0         0             0      No   
3             0         0           0         0             0      No   
4             1         1           0         0             0      No   

   waiting_time  Waiting_Time Weekday  
0             0             0  Friday  
1             0             0  Friday  
2             0             0  Friday  
3             0             0  Friday  
4             0             0  Friday

<a id='eda'></a>
## Exploratory Data Analysis

> **Tip**: Now that you've trimmed and cleaned your data, you're ready to move on to exploration. Compute statistics and create visualizations with the goal of addressing the research questions that you posed in the Introduction section. It is recommended that you be systematic with your approach. Look at one variable at a time, and then follow it up by looking at relationships between variables.

### Research Question 1 (Replace this header name!)

In [ ]:
# Use this, and more code cells, to explore your data. Don't forget to add
#   Markdown cells to document your observations and findings.


### Research Question 2  (Replace this header name!)

In [ ]:
# Continue to explore the data to address your additional research
#   questions. Add more headers as needed if you have more questions to
#   investigate.


<a id='conclusions'></a>
## Conclusions

> **Tip**: Finally, summarize your findings and the results that have been performed. Make sure that you are clear with regards to the limitations of your exploration. If you haven't done any statistical tests, do not imply any statistical conclusions. And make sure you avoid implying causation from correlation!

> **Tip**: Once you are satisfied with your work, you should save a copy of the report in HTML or PDF form via the **File** > **Download as** submenu. Before exporting your report, check over it to make sure that the flow of the report is complete. You should probably remove all of the "Tip" quotes like this one so that the presentation is as tidy as possible. Congratulations!